actors/miner/src/deadline_state.rs

```diff
@@ -99,6 +103,127 @@ impl Deadlines {
+
+        let first_to_partition_idx = to_partitions.count();
+        for (i, from_partition_idx) in partitions.iter().enumerate() {
+            let mut moving_partition = from_partitions
+                .get(from_partition_idx)?
+                .ok_or_else(|| actor_error!(not_found, "no partition {}", from_partition_idx))?
+                .clone();
+            if !moving_partition.faults.is_empty() || !moving_partition.unproven.is_empty() {
+                return Err(actor_error!(forbidden, "partition with faults or unproven sectors are not allowed to move, partition_idx {}", from_partition_idx))?;
+            }
+
+            let to_partition_idx = first_to_partition_idx + i as u64;
+
+            let from_expirations_epochs: Array<ExpirationSet, _> =
+                Array::load(&moving_partition.expirations_epochs, store)?;
+
+            let from_early_terminations: Array<BitField, _> =
+                Array::load(&moving_partition.early_terminated, store)?;
+
+            let mut to_expirations_epochs = Array::<ExpirationSet, BS>::new_with_bit_width(
+                store,
+                PARTITION_EXPIRATION_AMT_BITWIDTH,
+            );
+            let mut to_early_terminations = Array::<BitField, BS>::new_with_bit_width(
+                store,
+                PARTITION_EARLY_TERMINATION_ARRAY_AMT_BITWIDTH,
+            );
+
+            // ajust epoch info of expirations_epochs and early_terminated within `Partition`
+            from_expirations_epochs.for_each(|from_epoch, expire_set| {
+                to_expirations_epochs.set(
+                    to_quant.quantize_up(from_epoch as ChainEpoch) as u64,
+                    expire_set.clone(),
+                )?;
+                Ok(())
+            })?;
+            from_early_terminations.for_each(|from_epoch, bitfield| {
+                to_early_terminations
+                    .set(to_quant.quantize_up(from_epoch as ChainEpoch) as u64, bitfield.clone())?;
+                Ok(())
+            })?;
+            moving_partition.expirations_epochs = to_expirations_epochs.flush()?;
+            moving_partition.early_terminated = to_early_terminations.flush()?;
+
+            let all_sectors = moving_partition.sectors.len();
+            let live_sectors = moving_partition.live_sectors().len();
+            let early_terminations = from_deadline.early_terminations.get(from_partition_idx);
+            if early_terminations != (from_early_terminations.count() > 0) {
+                return Err(actor_error!(illegal_state, "Deadline.early_terminations doesn't agree with Partition.early_terminated for partition {}", from_partition_idx))?;
+            }
+
+            // start updating from/to `Deadline` here
+
+            from_deadline.total_sectors -= all_sectors;
+            from_deadline.live_sectors -= live_sectors;
+            from_deadline.faulty_power -= &moving_partition.faulty_power;
```

Comment:
I think most of the code outside the block that is updating Partition could be factored out into methods on Deadline. We can then test these much more easily than by orchestrating actual moves.

```
// Removes a partition from the deadline, updating its metadata.
// An empty partition is left in its place so that partition indices don't change.
// The partition's expiration queue entries are (not?) removed.
fn remove_partition(idx: u32) -> Result<Partition, ActorError>;

// Adds a partition to the deadline at the next index, updating deadline metadata.
// The partition's expiration queue keys are added to the deadline's expiration queue.
fn add_partition(p: Partition) -> Result<(), ActorError>;
```

Reply:
But we don't want to update `Deadline.expirations_epochs` for each `Partition`, instead we want to do it once for all partitions being moved. It's semantically weird if this function updates all other fields except `expirations_epochs`.